### PR TITLE
 GPU Resources 

### PR DESCRIPTION
**Pull request by DickyChant**

### CODE REVIEW

File: content/Resources/GPU_Resources/CMS_Resources/ml_cern_ch.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# ml.cern.ch
+
+[ml.cern.ch](https://ml.cern.ch) is a `Kubeflow` based ML solution provided by CERN. 
+
+## `Kubeflow` 
+[`Kubeflow`](https://www.kubeflow.org/docs/started/introduction/) is a Kubernetes based ML toolkits aiming at making deployments of ML workflows simple, portable and scalable. In Kubeflow, *pipeline* is an important concept. Machine Learning workflows are discribed as a Kubeflow *pipeline* for execution.
+
+## How to access
+[`ml.cern.ch`](https://ml.cern.ch) only accepts connection from CERN internet. Therefore, if you are outside of CERN a network tunneling will be needed (e.g. via `ssh -D` dynamical port forwarding as proxy). The main website are shown below.
```
Comment (aperloff) [inline content/Resources/GPU_Resources/CMS_Resources/ml_cern_ch.md:9]: "only accepts connections from within the CERN network. Therefore, if you are outside of CERN, you will need to use a network tunnel (eg. via `ssh -D` dynamic port forwarding as proxy)..."

Is it really the `-D` option that you want of is it `-L` as is recommended [here](https://linuxize.com/post/how-to-setup-ssh-tunneling/)?

File: content/Resources/GPU_Resources/index.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# CMS GPU Resources
+
+# Facilities
```
Comment (aperloff) [inline content/Resources/GPU_Resources/index.md:3]: Remove this line

File: content/Resources/GPU_Resources/index.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+1. Collaboration-wide GPU resources
```
Comment (aperloff) [inline content/Resources/GPU_Resources/index.md:5]: I don't think you need to specify that these are collaboration wide resources. This is implied as there are no restricted resources listed right now. I would simply number the four sub-items.

**Review by aperloff — COMMENTED** [review]
A couple of general comments:
1. Rename your images with descriptive file names.
2. Since this documentation will be open to everyone on the internet, you may want to blur out or cover up your CERN username. It is visible in some of the images.

**Comment by DickyChant** [comment]
> A couple of general comments:
> 
> 1. Rename your images with descriptive file names.
> 2. Since this documentation will be open to everyone on the internet, you may want to blur out or cover up your CERN username. It is visible in some of the images.

Thanks for these two points, I have changed the filenames and blurred my username.

File: mkdocs.yml [hunk 4]
```diff
@@ -117,3 +117,7 @@ nav:
       - Successful integrations:
         - ParticleNet: inference/particlenet.md
 #    - Benchmarking:
+  - Resources:
+    - GPU Resources: Resources/GPU_Resources/index.md
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply mkdocs.yml]: They are arranged in alphabetical order now.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+2. Setup SWAN with GPU resources.
+    1. Once the registration is done, one can login [SWAN with Kerberes8 support](https://swan-k8s.cern.ch/) and then create his SWAN environment.
+        
+        <aside>
+        💡 Note. When configuring SWAN environment, one should carefully choose the software stack he want to use. Not only one should choose to use the releases with GPU Support, but also she/he should keep the CUDA version in mind. Once additional software installation is needed, it must be configured to use the version compatible with current CUDA version.
+        
+        </aside>
+        
+
+![Untitled](SWAN_figs/Untitled.png)
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: OK, all figure names changed.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# SWAN
+## Preparation
+
+1. Registration:
+    To require GPU resources for SWAN: According to [this thread](https://swan-community.web.cern.ch/t/gpu-support-in-swan/108), one can create a ticket through [this link](https://cern.service-now.com/service-portal?id=functional_element&name=swan) to ask for GPU support at SWAN, it is now in beta version and limited to a small scale.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md:6]: Thanks for the reminder, I have put a blank line to make "To require..." on a new line.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+2. Setup SWAN with GPU resources.
+    1. Once the registration is done, one can login [SWAN with Kerberes8 support](https://swan-k8s.cern.ch/) and then create his SWAN environment.
+        
+        <aside>
+        💡 Note. When configuring SWAN environment, one should carefully choose the software stack he want to use. Not only one should choose to use the releases with GPU Support, but also she/he should keep the CUDA version in mind. Once additional software installation is needed, it must be configured to use the version compatible with current CUDA version.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Changed

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        
+        </aside>
+        
+
+![Untitled](SWAN_figs/Untitled.png)
+
+![Untitled](SWAN_figs/Untitled%201.png)
+
+Another important aspects is the environment script, which will be discussed later in this document.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Changed

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+1. After setting up the SWAN environment, one will browse the SWAN main directory `My Project` where all existing projects are displayed. A new project can be created by clicking the upper right "+" button. After creation one will be redirected to the new created project, whose "+" button at upper right panel can be used for creating new **notebook**.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Thanks for your suggestion, I have updated this 2nd level header.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    
+    ![Untitled](SWAN_figs/Untitled%202.png)
+    
+    ![Untitled](SWAN_figs/Untitled%203.png)
+    
+2. Also it is possible to use the terminal for installing new packages or monitoring computational resources.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Thanks for the reminder.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+Another important aspects is the environment script, which will be discussed later in this document.
+
+1. After setting up the SWAN environment, one will browse the SWAN main directory `My Project` where all existing projects are displayed. A new project can be created by clicking the upper right "+" button. After creation one will be redirected to the new created project, whose "+" button at upper right panel can be used for creating new **notebook**.
+    
+    ![Untitled](SWAN_figs/Untitled%202.png)
+    
+    ![Untitled](SWAN_figs/Untitled%203.png)
+    
+2. Also it is possible to use the terminal for installing new packages or monitoring computational resources. 
+    1. For package installation, one can install packages with package management tools, e.g. `pip` for `python`. To use the installed package, one need to wrap environment configuration into a script and let SWAN to execute. Detailed documentation can be found by clicking the upper right "?" button.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Thanks for the rephrasing.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    2. To monitor the computational resources, in addition to ordinary CPU resources `top` and `htop` , you can also use the `nvidia-smi` to monitor GPU usage.
+    
+    ![Untitled](SWAN_figs/Untitled%204.png)
+    
+
+## Examples 
+
+After installing package, you can then use GPU based machine learning algorithms. Two examples are supplied as an example.
+
+> 1). The first example aims at using a CNN to perform handwritten digits classification with `MNIST` dataset. The whole notebook can be found at [PytorchMNIST](Notebooks/PytorchMNIST.md). This example is modified from [an official `pytorch` example](https://github.com/pytorch/examples/tree/master/mnist).
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Thanks for the reminder, the syntax has been updated in order to form a list.

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_gpu.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
+# lxplus-gpu.cern.ch
+
+## How to use it?
+
+`lxplus-gpu` are special lxplus nodes with GPU support. Therefore, via executing
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_gpu.md]: Changed

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_gpu.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+ssh <your_user_name>@lxplus-gpu.cern.ch
+```
+
+one can log in to such a node equipped with GPU support.
+
+![Untitled](LxplusGPU_figs/Untitled.png)
+
+## Software Environment
+
+Three examples are given below to show how to set up a software environment properly.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_gpu.md]: Sorry for this mistake😅

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_gpu.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        # Execute the command shown in your terminal
+        pip3 install torch==1.10.0+cu113 torchvision==0.11.1+cu113 torchaudio==0.10.0+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html
+        pip3 install jupyterlab matplotlib scikit-hep # install other packages if they are needed
+        ```
+3. Using `conda` package manager: `conda` pacakge manager is more convenient to install and use. To begin with, obtaining an `Anaconda` or `Miniconda` installer for Linux x86_64 platform. Then execute it on Lxplus.
+
+    1. Please note that if you update your shell configuration (e.g. `.bashrc` file) by `conda init`, you may encounter failure due to inconsistent environment configuration.
+    2. Installing packages via `conda` also needs special consideration on selecting proper CUDA version as discussed in `pip` part.
+
+4. Container based solution: There exists several container images in `unpacked.cern.ch` inside `CVMFS` which often contain a large set of softwares and thus provide a complete computing environment. If you find a image helpful, you can use `singularity` command to load and use it.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_gpu.md]: Thanks for providing a singularity command!

File: content/Resources/GPU_Resources/index.md (hunk 2 above)
Comment (DickyChant) [reply content/Resources/GPU_Resources/index.md]: Removed

File: content/Resources/GPU_Resources/index.md (hunk 3 above)
Comment (DickyChant) [reply content/Resources/GPU_Resources/index.md]: Thanks for your comment, I use a numbered list instead.

File: content/Resources/GPU_Resources/CMS_Resources/ml_cern_ch.md (hunk 1 above)
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/ml_cern_ch.md]: Thanks for the rephrase. Here the `-D` option is used to create a SOCKS5 proxy so one can browse CERN internet through it. With a SOCKS5 proxy, the data will be viewed as within CERN internet.

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+2. A paradigm example on submitting GPU HTCondor job @ Lxplus is the [`weaver-benchmark`](https://github.com/colizz/weaver-benchmark) project. It provides a concrete example on how to setup environment for `weaver` framework and operate trainning and testing process within a single job. Detailed description can be found at section [`ParticleNet`](../../../inference/particlenet.md) of this documentation.
+3. CMS Connect also provides a [documentation](https://ci-connect.atlassian.net/wiki/spaces/CMS/pages/80117822/Requesting+GPUs) on GPU job submission. In this documentation there is also a `Tensorflow` example. 
+   
+        💡 Note. When submitting GPU jobs @ CMS Connect, especially for Machine Learning purpose, EOS space @ CERN are not accessible as directory, therefore one should consider using `xrootd` service as documented in [this page](https://ci-connect.atlassian.net/wiki/spaces/CMS/pages/850264068/Using+stashcp+and+XrootD+from+OSG+TensorFlow+containers)
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md]: Thanks for the rephrasing.

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+People can require their jobs to have GPU support by adding the following requirements to the condor submission file.
+
+>```bash
+>request_gpus = n # n equal to the number of GPUs required
+
+## Further documentation
+
+There are good materials providing detailed documentation on how to run HTCondor jobs with GPU support at both machines. 
+1. A complete documentation can be found from the `GPUs` section in [CERN Batch Docs](https://batchdocs.web.cern.ch/tutorial/exercise10.html). Where a `Tensorflow` example is supplied. This documentation also contains instructions on advanced HTCondor configuration, for instance constraining GPU device or CUDA version.
+2. A paradigm example on submitting GPU HTCondor job @ Lxplus is the [`weaver-benchmark`](https://github.com/colizz/weaver-benchmark) project. It provides a concrete example on how to setup environment for `weaver` framework and operate trainning and testing process within a single job. Detailed description can be found at section [`ParticleNet`](../../../inference/particlenet.md) of this documentation.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md]: This example can be run elsewhere, I have mentioned this in the updated version.

File: content/Resources/GPU_Resources/CMS_Resources/SWAN.md [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+1. After setting up the SWAN environment, one will browse the SWAN main directory `My Project` where all existing projects are displayed. A new project can be created by clicking the upper right "+" button. After creation one will be redirected to the new created project, whose "+" button at upper right panel can be used for creating new **notebook**.
+    
+    ![Untitled](SWAN_figs/Untitled%202.png)
+    
+    ![Untitled](SWAN_figs/Untitled%203.png)
+    
+2. Also it is possible to use the terminal for installing new packages or monitoring computational resources. 
+    1. For package installation, one can install packages with package management tools, e.g. `pip` for `python`. To use the installed package, one need to wrap environment configuration into a script and let SWAN to execute. Detailed documentation can be found by clicking the upper right "?" button.
+    2. To monitor the computational resources, in addition to ordinary CPU resources `top` and `htop` , you can also use the `nvidia-smi` to monitor GPU usage.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/SWAN.md]: Thanks for the rephrasing.

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
+# lxplus condor
+
+The `HTCondor` service @ `Lxplus` has GPU support. Therefore one can submit jobs on machine learning tasks to make use of this resources.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md]: Good point, this part is something I should have deleted. I have removed this paragraph in the updated version.

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+A complete documentation can be found from the `GPUs` section in [CERN Batch Docs](https://batchdocs.web.cern.ch/tutorial/exercise10.html). Where a `Tensorflow` example is supplied. For `pytorch` users, it is worthwhile to take a look at [`weaver-benchmark`](https://github.com/colizz/weaver-benchmark). This repository not only contains tools for benchmarking, but also offers a good example on how to use `HTCondor` @ `Lxplus` to train `pytorch` ML models with `weaver`. Detailed explanation and tutorial on how to use [`weaver-benchmark`](https://github.com/colizz/weaver-benchmark) can be found at section [`ParticleNet`](../../../inference/particlenet.md) of this documentation.
+
+# HTCondor With GPU resources
+In general, HTCondor supports GPU jobs if there are some GPU devices configured. For two main collaboration-wide HTCondor facilities, Lxplus and CMS Connect, they both have GPU resources equipped. 
+
+## How to require GPUs in HTCondor
+
+People can require their jobs to have GPU support by adding the following requirements to the condor submission file.
+
+>```bash
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md]: Thanks, now I use ` ``` ` only.

File: content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+>request_gpus = n # n equal to the number of GPUs required
+
+## Further documentation
+
+There are good materials providing detailed documentation on how to run HTCondor jobs with GPU support at both machines. 
+1. A complete documentation can be found from the `GPUs` section in [CERN Batch Docs](https://batchdocs.web.cern.ch/tutorial/exercise10.html). Where a `Tensorflow` example is supplied. This documentation also contains instructions on advanced HTCondor configuration, for instance constraining GPU device or CUDA version.
```
[the remark this answers is not in the document]
Comment (DickyChant) [reply content/Resources/GPU_Resources/CMS_Resources/lxplus_condor.md:15]: Thanks for the reminder.

**Comment by aperloff** [comment]
Looks good to me. If you are also happy, remove the `[WIP]` notation from the title and I will merge it in.

**Comment by DickyChant** [comment]
Thanks, let me modify the title.

**Pull request merged**